### PR TITLE
Fix bug in `remove_join_expressions`

### LINKED ISSUE #11621: Incorrect predicate evaluation result in a query (SQLancer-NoREC)
### Describe the bug

In the following reproducer, two rows in each table should first be joined, then get evaluted to `true` by the predicate, and finally output to the final result.
But the final result showed 0 rows output, it should be a logic bug.
(Verified in DuckDB and PostgreSQL)

### To Reproduce

Run in datafusion-cli
```
    /*DML*/CREATE TABLE t1(v0 DOUBLE, v1 BOOLEAN, v2 STRING, v3 STRING);
    /*DML*/CREATE TABLE t48(v0 DOUBLE, v1 STRING);
	
    /*DML*/INSERT INTO t1(v3, v0) VALUES ('', '-Inf'::Double);
    /*DML*/INSERT INTO t48(v1, v0) VALUES ('', '-Inf'::Double);
```
And the executed result
```
> SELECT * FROM t1 JOIN t48 ON ((t1.v0)=(t48.v0)) WHERE ((((t48.v0)==(t1.v0)))OR(t1.v1));
+----+----+----+----+----+----+
| v0 | v1 | v2 | v3 | v0 | v1 |
+----+----+----+----+----+----+
+----+----+----+----+----+----+
0 row(s) fetched.
Elapsed 0.041 seconds.
```
expected result should be 1 row output:
```
D SELECT * FROM t1 JOIN t48 ON ((t1.v0)=(t48.v0)) WHERE ((((t48.v0)==(t1.v0)))OR(t1.v1));
┌────────┬─────────┬─────────┬─────────┬────────┬─────────┐
│   v0   │   v1    │   v2    │   v3    │   v0   │   v1    │
│ double │ boolean │ varchar │ varchar │ double │ varchar │
├────────┼─────────┼─────────┼─────────┼────────┼─────────┤
│   -inf │         │         │         │   -inf │         │
└────────┴─────────┴─────────┴─────────┴────────┴─────────┘
```

### Expected behavior

_No response_

### Additional context

Found by SQLancer https://github.com/apache/datafusion/issues/11030

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?

Closes #11621.

## Rationale for this change
Given an example query: 
```sql
SELECT * FROM t1 JOIN t2 ON t1.v1 = t2.v1 WHERE (t1.v1 == t2.v1) OR t1.v1
```
When removing the join predicate `t1.v1 = t2.v1` from the parent filter `(t1.v1 == t2.v1) OR t1.v1`,  this process can be considered as replacing the join predicate with `true`. The result should be `TRUE OR t1.v1` and further simplified to empty, but currently, it is `t1.v1`. This will lead to incorrect join results.

## Are these changes tested?
Yes

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 1]
```diff
@@ -402,7 +400,18 @@ fn remove_join_expressions(expr: Expr, join_keys: &JoinKeySet) -> Option<Expr> {
...
+        Expr::BinaryExpr(BinaryExpr { left, op, right }) if op == Operator::Or => {
+            let l = remove_join_expressions(*left, join_keys);
+            let r = remove_join_expressions(*right, join_keys);
+            match (l, r) {
+                (Some(ll), Some(rr)) => Some(Expr::BinaryExpr(BinaryExpr::new(
+                    Box::new(ll),
+                    op,
+                    Box::new(rr),
+                ))),
+                _ => None,
```
Comment (jonahgao) [inline datafusion/optimizer/src/eliminate_cross_join.rs:412]: When either `left` or `right` is empty, it means they are `true`.

File: datafusion/sqllogictest/test_files/join.slt [hunk 2]
```diff
@@ -1046,6 +1046,19 @@ true false
 true NULL
 true NULL
 
+# Test issue: https://github.com/apache/datafusion/issues/11621
+query BB
+SELECT * FROM t1 JOIN t2 ON t1.v1 = t2.v1 WHERE (t1.v1 == t2.v1) OR t1.v1;
```
Comment (jonahgao) [inline datafusion/sqllogictest/test_files/join.slt:1051]: This query returns an empty result on the current main.

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 3]
```diff
@@ -1057,7 +1071,7 @@ mod tests {
             "Filter: (t4.c < UInt32(15) OR t4.c = UInt32(688)) AND (t4.c < UInt32(15) OR t3.c = UInt32(688) OR t3.b = t4.b) [a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32]",
             "  Inner Join: t3.a = t4.a [a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32]",
             "    Inner Join: t1.a = t3.a [a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32]",
-            "      Filter: t2.c < UInt32(15) AND t2.c = UInt32(688) [a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32]",
+            "      Filter: t2.c = UInt32(688) [a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32]",
```
Comment (jonahgao) [inline datafusion/optimizer/src/eliminate_cross_join.rs:1074]: Remove `t1.a = t2.a` from `(t1.a = t2.a OR t2.c < 15) AND (t1.a = t2.a AND tc.2 = 688)`, and the result should be ` tc.2 = 688`

**Comment by jonahgao** [comment]
The failure of clippy is unrelated to the current changes.
> error: use of deprecated method `arrow_schema::Schema::all_fields`: Use `flattened_fields` instead

File: datafusion/optimizer/src/eliminate_cross_join.rs (hunk 1 above)
Comment (alamb) [reply datafusion/optimizer/src/eliminate_cross_join.rs:412]: That might be useful to encode as a comment:
```suggestion
               // When either `left` or `right` is empty, it means they are `true`
               // so OR'ing anything with them will also be true
                _ => None,
```

**Review by alamb — APPROVED** [review]
This makes sense to me. Thank you @jonahgao for the fix (and @2010YOUY01  for SQLancer that found it 🙏 )

**Comment by jonahgao** [comment]
Thanks for the review @alamb

**Pull request merged**